### PR TITLE
Feature 11 retrieve sap contracts

### LINKED ISSUE #11: Retrieve SAP contracts

### PR DESCRIPTION
**Pull request by platonovsimeon**
Resolves #11

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/ContractsController.kt [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        val groupCodes = configController.getGroupCodesFile()
+        val itemPropertiesSelect = getItemPropertiesSelect(groupCodes = groupCodes)
+        val items = getItemsAsJsonNodes(
+            resourceUrl = "${session.apiUrl}/Items",
+            select = itemPropertiesSelect,
+            sessionId = session.sessionId,
+            routeId = session.routeId
+        )
+
+        contracts.forEach { contract ->
```
Comment (jarinykanen) [inline src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/ContractsController.kt:87]: Split multiple for each loops to own funtions

File: src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+     * Lists all contracts in SAP, from beginning of the year 2022
+     *
+     * @param sapContractStatus SAP contract status to filter by
+     * @return SAP contracts
+     */
+    fun list(sapContractStatus: SapContractStatus?): Array<SapContract> {
+        return api.listContracts(startDate = null, businessPartnerCode = null, contractStatus = sapContractStatus)
+    }
+
+    fun create(sapContract: SapContract): SapContract {
```
Comment (jarinykanen) [inline src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt:33]: Add docs

File: src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    override fun getApi(): ContractsApi {
+        ApiClient.accessToken = accessTokenProvider?.accessToken
+        return ContractsApi(ApiTestSettings.apiBasePath)
+    }
+
+    fun assertListFailStatus(
+        expectedStatus: Int,
+    ) {
```
Comment (jarinykanen) [inline src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt:44]:
```suggestion
    fun assertListFailStatus(expectedStatus: Int) {
```

Comment (jarinykanen) [reply src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt:44]: Add docs

**Review by jarinykanen — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/jaxrs/LocalDateParamProvider.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package fi.metatavu.pakkasmarja.services.erp.jaxrs
+
+import java.lang.reflect.Type
+import java.time.LocalDate
+import javax.ws.rs.ext.ParamConverter
+import javax.ws.rs.ext.ParamConverterProvider
+import javax.ws.rs.ext.Provider
+
+@Provider
+class LocalDateParamProvider: ParamConverterProvider {
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/jaxrs/LocalDateParamProvider.kt:10]: Added

File: src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/AbstractSapResourceController.kt [hunk 5]
```diff
@@ -62,34 +157,34 @@ abstract class AbstractSapResourceController {
...
-                    val items = objectMapper.readTree(response.body()).get("value")
-                    items.forEach { item ->
-                        jsonNodes.add(item)
+        try {
+            val jsonNodes = mutableListOf<JsonNode>()
+            val client = HttpClient.newHttpClient()
+            val futures = mutableListOf<CompletableFuture<Unit>>()
+
+            itemUrls.forEach {
+                    val request = HttpRequest
```
[the remark this answers is not in the document]
Comment (platonovsimeon) [reply src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/AbstractSapResourceController.kt]: Fixed

File: src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt (hunk 2 above)
Comment (platonovsimeon) [reply src/test/kotlin/fi/metatavu/pakkasmarja/services/erp/test/functional/impl/ContractsResourceTestBuilderResource.kt:39]: Fixed

File: src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/ContractsController.kt (hunk 1 above)
Comment (platonovsimeon) [reply src/main/kotlin/fi/metatavu/pakkasmarja/services/erp/sap/ContractsController.kt:92]: Splitted

**Review by jarinykanen — APPROVED** [review]

**Pull request merged**